pkg/server/framework/server.go

```diff
@@ -19,78 +18,82 @@ import (
 
 const (
-	KeyRequestState ctxKey = 1
-	serviceName            = "ssi-service"
+	KeyRequestState    contextKey = "keyRequestState"
+	ShutdownErrorState contextKey = "shutdownError"
+
+	serviceName string = "ssi-service"
 )
 
+func (c contextKey) String() string {
+	return string(c)
+}
+
 type RequestState struct {
 	TraceID    string
 	Now        time.Time
 	StatusCode int
 }
 
-// A Handler is a type that handles a http request within our own little mini
-// framework.
-type Handler func(ctx context.Context, w http.ResponseWriter, r *http.Request) error
-
-// Server is the entrypoint into our application and what configures our context
-// object for each of our http router. Feel free to add any configuration
-// data/logic on this Server struct.
+// Server is the entrypoint into our application and what configures our context object for each of our http router.
+// Feel free to add any configuration data/logic on this Server struct.
 type Server struct {
-	*httptreemux.ContextMux
+	*http.Server
+	router   *gin.Engine
 	tracer   trace.Tracer
 	shutdown chan os.Signal
-	mw       []Middleware
 }
 
+type Handler func(c *gin.Context) error
+
 // NewHTTPServer creates a Server that handles a set of routes for the application.
-func NewHTTPServer(config config.ServerConfig, shutdown chan os.Signal, mw ...Middleware) *Server {
+func NewHTTPServer(config config.ServerConfig, shutdown chan os.Signal, mws gin.HandlersChain) *Server {
 	var tracer trace.Tracer
 	if config.JagerEnabled {
 		tracer = otel.Tracer(serviceName)
 	}
+	router := gin.Default()
+	router.Use(mws...)
+
 	return &Server{
-		ContextMux: httptreemux.NewContextMux(),
-		tracer:     tracer,
-		shutdown:   shutdown,
-		mw:         mw,
+		Server: &http.Server{
+			Addr:              config.APIHost,
+			Handler:           router,
+			ReadTimeout:       config.ReadTimeout,
+			ReadHeaderTimeout: config.ReadTimeout,
+			WriteTimeout:      config.WriteTimeout,
+		},
+		router:   router,
+		tracer:   tracer,
+		shutdown: shutdown,
 	}
 }
 
-func (s *Server) AddMiddleware(mw Middleware) {
-	s.mw = append(s.mw, mw)
-}
-
 // Handle sets a handler function for a given HTTP method and path pair
 // to the server mux.
-func (s *Server) Handle(method string, path string, handler Handler, mw ...Middleware) {
-	// first wrap route specific middleware
-	handler = WrapMiddleware(mw, handler)
-
-	// then wrap app specific middleware
-	handler = WrapMiddleware(s.mw, handler)
+func (s *Server) Handle(method string, path string, handler Handler, middleware ...gin.HandlerFunc) {
+	// add the middleware to the router
+	s.router.Use(middleware...)
 
 	// request handler function
-	h := func(w http.ResponseWriter, r *http.Request) {
+	h := func(c *gin.Context) {
 		requestState := RequestState{
 			TraceID: uuid.New().String(),
 			Now:     time.Now(),
 		}
-		ctx := context.WithValue(r.Context(), KeyRequestState, &requestState)
+		r := c.Request
+		c.Set(KeyRequestState.String(), &requestState)
```

Comment:
We shouldn't be creating our own traces. This should be handled by otel (since traces may come from the request). Maybe for an issue.

Reply:
yeah, this predates me, was something Moe had added. I'm open to removing it.

Reply:
will be handled in #439